File: apps/api/momento_api/routes/albums.py
from typing import Annotated

from fastapi import APIRouter, Depends, HTTPException, status

from momento_api.auth.dependencies import CurrentUser, get_current_user
from momento_api.database import execute_query, fetch_all, fetch_one, get_connection, insert_returning_id
from momento_api.models.album import (
    AlbumAddMediaRequest,
    AlbumCreateRequest,
    AlbumDeleteRequest,
    AlbumDetailResponse,
    AlbumGetRequest,
    AlbumListResponse,
    AlbumRemoveMediaRequest,
    AlbumReorderRequest,
    AlbumResponse,
    AlbumUpdateRequest,
)
from momento_api.models.media import MediaResponse

router = APIRouter(prefix="/album", tags=["album"])


def _row_to_album_response(row) -> AlbumResponse:
    return AlbumResponse(
        id=row["id"],
        name=row["name"],
        description=row["description"],
        cover_media_id=row["cover_media_id"],
        media_count=row["media_count"] if "media_count" in row.keys() else 0,
        created_at=row["created_at"],
    )


def _row_to_media_response(row) -> MediaResponse:
    return MediaResponse(
        id=row["id"],
        filename=row["filename"],
        original_filename=row["original_filename"],
        media_type=row["media_type"],
        mime_type=row["mime_type"],
        width=row["width"],
        height=row["height"],
        file_size=row["file_size"],
        duration_seconds=row["duration_seconds"],
        date_taken=row["date_taken"],
        gps_latitude=row["gps_latitude"],
        gps_longitude=row["gps_longitude"],
        camera_make=row["camera_make"],
        camera_model=row["camera_model"],
        iso=row["iso"],
        exposure_time=row["exposure_time"],
        f_number=row["f_number"],
        focal_length=row["focal_length"],
        gps_altitude=row["gps_altitude"],
        location_state=row["location_state"],
        location_country=row["location_country"],
        keywords=row["keywords"],
        created_at=row["created_at"],
    )


@router.post("/create", response_model=AlbumResponse)
async def create_album(
    request: AlbumCreateRequest, current_user: Annotated[CurrentUser, Depends(get_current_user)]
) -> AlbumResponse:
    album_id = insert_returning_id(
        "INSERT INTO albums (user_id, name, description) VALUES (?, ?, ?)",
        (current_user.id, request.name, request.description),
    )

    row = fetch_one("SELECT a.*, 0 as media_count FROM albums a WHERE a.id = ?", (album_id,))
    if row is None:
        raise HTTPException(status_code=status.HTTP_500_INTERNAL_SERVER_ERROR, detail="Failed to create album")

    return _row_to_album_response(row)


@router.post("/list", response_model=AlbumListResponse)
async def list_albums(current_user: Annotated[CurrentUser, Depends(get_current_user)]) -> AlbumListResponse:
    rows = fetch_all(
        """
        SELECT a.*, COUNT(am.media_id) as media_count
        FROM albums a
        LEFT JOIN album_media am ON a.id = am.album_id
        WHERE a.user_id = ?
        GROUP BY a.id
        ORDER BY a.created_at DESC
        """,
        (current_user.id,),
    )
    return AlbumListResponse(albums=[_row_to_album_response(row) for row in rows])


@router.post("/get", response_model=AlbumDetailResponse)
async def get_album(
    request: AlbumGetRequest, current_user: Annotated[CurrentUser, Depends(get_current_user)]
) -> AlbumDetailResponse:
    album_row = fetch_one("SELECT * FROM albums WHERE id = ? AND user_id = ?", (request.album_id, current_user.id))
    if album_row is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Album not found")

    media_rows = fetch_all(
        """
        SELECT m.* FROM media m
        JOIN album_media am ON m.id = am.media_id
        WHERE am.album_id = ?
        ORDER BY am.position
        """,
        (request.album_id,),
    )

    return AlbumDetailResponse(
        id=album_row["id"],
        name=album_row["name"],
        description=album_row["description"],
        cover_media_id=album_row["cover_media_id"],
        media=[_row_to_media_response(row) for row in media_rows],
        created_at=album_row["created_at"],
    )


@router.post("/update", response_model=AlbumResponse)
async def update_album(
    request: AlbumUpdateRequest, current_user: Annotated[CurrentUser, Depends(get_current_user)]
) -> AlbumResponse:
    album_row = fetch_one("SELECT * FROM albums WHERE id = ? AND user_id = ?", (request.album_id, current_user.id))
    if album_row is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Album not found")

    updates: list[str] = []
    params: list[object] = []

    if request.name is not None:
        updates.append("name = ?")
        params.append(request.name)

    if request.description is not None:
        updates.append("description = ?")
        params.append(request.description)

    if request.cover_media_id is not None:
        updates.append("cover_media_id = ?")
        params.append(request.cover_media_id)

    if updates:
        params.append(request.album_id)
        execute_query(f"UPDATE albums SET {', '.join(updates)} WHERE id = ?", tuple(params))
        get_connection().commit()

    row = fetch_one(
        """
        SELECT a.*, COUNT(am.media_id) as media_count
        FROM albums a
        LEFT JOIN album_media am ON a.id = am.album_id
        WHERE a.id = ?
        GROUP BY a.id
        """,
        (request.album_id,),
    )
    return _row_to_album_response(row)


@router.post("/delete")
async def delete_album(
    request: AlbumDeleteRequest, current_user: Annotated[CurrentUser, Depends(get_current_user)]
) -> dict:
    album_row = fetch_one("SELECT id FROM albums WHERE id = ? AND user_id = ?", (request.album_id, current_user.id))
    if album_row is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Album not found")

    execute_query("DELETE FROM albums WHERE id = ?", (request.album_id,))
    get_connection().commit()

    return {"message": "Album deleted successfully"}


@router.post("/add-media")
async def add_media_to_album(
    request: AlbumAddMediaRequest, current_user: Annotated[CurrentUser, Depends(get_current_user)]
) -> dict:
    album_row = fetch_one("SELECT id FROM albums WHERE id = ? AND user_id = ?", (request.album_id, current_user.id))
    if album_row is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Album not found")

    max_pos_row = fetch_one(
        "SELECT COALESCE(MAX(position), -1) as max_pos FROM album_media WHERE album_id = ?", (request.album_id,)
    )
    next_pos = (max_pos_row["max_pos"] if max_pos_row else -1) + 1

    conn = get_connection()
    insert_params: list[tuple[int, int, int]] = []
    for index, media_id in enumerate(request.media_ids):
        media_row = fetch_one("SELECT id FROM media WHERE id = ? AND user_id = ?", (media_id, current_user.id))
        if media_row is None:
            continue
        insert_params.append((request.album_id, media_id, next_pos + index))

    if not insert_params:
        return {"message": "Media added to album"}

    conn.executemany("INSERT OR IGNORE INTO album_media (album_id, media_id, position) VALUES (?, ?, ?)", insert_params)
    conn.commit()

    return {"message": "Media added to album"}


@router.post("/remove-media")
async def remove_media_from_album(
    request: AlbumRemoveMediaRequest, current_user: Annotated[CurrentUser, Depends(get_current_user)]
) -> dict:
    album_row = fetch_one("SELECT id FROM albums WHERE id = ? AND user_id = ?", (request.album_id, current_user.id))
    if album_row is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Album not found")

    conn = get_connection()
    for media_id in request.media_ids:
        conn.execute("DELETE FROM album_media WHERE album_id = ? AND media_id = ?", (request.album_id, media_id))
    conn.commit()

    return {"message": "Media removed from album"}


@router.post("/reorder")
async def reorder_album_media(
    request: AlbumReorderRequest, current_user: Annotated[CurrentUser, Depends(get_current_user)]
) -> dict:
    album_row = fetch_one("SELECT id FROM albums WHERE id = ? AND user_id = ?", (request.album_id, current_user.id))
    if album_row is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Album not found")

    conn = get_connection()
    for i, media_id in enumerate(request.media_ids):
        conn.execute(
            "UPDATE album_media SET position = ? WHERE album_id = ? AND media_id = ?", (i, request.album_id, media_id)
        )
    conn.commit()

    return {"message": "Album reordered successfully"}

File: src/api/src/routes/albums.rs
use axum::{extract::State, routing::post, Json, Router};

use crate::auth::{AppState, CurrentUser};
use crate::database::{execute_query, fetch_all, fetch_one, insert_returning_id, queries};
use crate::error::{AppError, AppResult};
use crate::models::{
    AlbumAddMediaRequest, AlbumCreateRequest, AlbumDeleteRequest, AlbumDetailResponse,
    AlbumGetRequest, AlbumListResponse, AlbumRemoveMediaRequest, AlbumReorderRequest,
    AlbumResponse, AlbumUpdateRequest, MediaResponse,
};

pub fn router() -> Router<AppState> {
    Router::new()
        .route("/album/create", post(create_album))
        .route("/album/list", post(list_albums))
        .route("/album/get", post(get_album))
        .route("/album/update", post(update_album))
        .route("/album/delete", post(delete_album))
        .route("/album/add-media", post(add_media_to_album))
        .route("/album/remove-media", post(remove_media_from_album))
        .route("/album/reorder", post(reorder_album_media))
}

fn map_album_row(row: &rusqlite::Row) -> rusqlite::Result<AlbumResponse> {
    Ok(AlbumResponse {
        id: row.get(0)?,
        name: row.get(1)?,
        description: row.get(2)?,
        cover_media_id: row.get(3)?,
        media_count: row.get(4)?,
        created_at: row.get(5)?,
    })
}

async fn create_album(
    State(state): State<AppState>,
    current_user: CurrentUser,
    Json(request): Json<AlbumCreateRequest>,
) -> AppResult<Json<AlbumDetailResponse>> {
    let conn = state.pool.get().map_err(AppError::Pool)?;

    let album_id = insert_returning_id(
        &conn,
        queries::albums::INSERT,
        &[&current_user.id, &request.name, &request.description],
    )?;

    execute_query(
        &conn,
        queries::access::INSERT_ALBUM_ACCESS,
        &[&album_id, &current_user.id, &2],
    )?;

    let album = fetch_one(&conn, queries::albums::SELECT_BY_ID, &[&album_id], |row| {
        Ok(AlbumBasic {
            id: row.get(0)?,
            name: row.get(1)?,
            description: row.get(2)?,
            cover_media_id: row.get(3)?,
            created_at: row.get(5)?,
        })
    })?
    .ok_or_else(|| AppError::NotFound("Album not found".to_string()))?;

    let media = fetch_all(
        &conn,
        queries::albums::SELECT_MEDIA,
        &[&album_id],
        map_media_row,
    )?;

    Ok(Json(AlbumDetailResponse {
        id: album.id,
        name: album.name,
        description: album.description,
        cover_media_id: album.cover_media_id,
        media,
        created_at: album.created_at,
    }))
}

struct AlbumBasic {
    id: i64,
    name: String,
    description: Option<String>,
    cover_media_id: Option<i64>,
    created_at: String,
}

fn map_media_row(row: &rusqlite::Row) -> rusqlite::Result<MediaResponse> {
    Ok(MediaResponse {
        id: row.get(0)?,
        filename: row.get(1)?,
        original_filename: row.get(2)?,
        media_type: row.get(3)?,
        mime_type: row.get(4)?,
        width: row.get(5)?,
        height: row.get(6)?,
        file_size: row.get(7)?,
        duration_seconds: row.get(8)?,
        date_taken: row.get(9)?,
        gps_latitude: row.get(10)?,
        gps_longitude: row.get(11)?,
        camera_make: row.get(12)?,
        camera_model: row.get(13)?,
        lens_make: row.get(14)?,
        lens_model: row.get(15)?,
        iso: row.get(16)?,
        exposure_time: row.get(17)?,
        f_number: row.get(18)?,
        focal_length: row.get(19)?,
        focal_length_35mm: row.get(20)?,
        gps_altitude: row.get(21)?,
        location_city: row.get(22)?,
        location_state: row.get(23)?,
        location_country: row.get(24)?,
        video_codec: row.get(25)?,
        keywords: row.get(26)?,
        created_at: row.get(27)?,
        content_hash: None,
    })
}

async fn update_album(
    State(state): State<AppState>,
    current_user: CurrentUser,
    Json(request): Json<AlbumUpdateRequest>,
) -> AppResult<Json<AlbumResponse>> {
    let conn = state.pool.get().map_err(AppError::Pool)?;

    let exists = fetch_one(
        &conn,
        queries::albums::CHECK_OWNERSHIP,
        &[&request.album_id, &current_user.id],
        |row| row.get::<_, i64>(0),
    )?;

    if exists.is_none() {
        return Err(AppError::NotFound("Album not found".to_string()));
    }

    let mut updates = Vec::new();
    let mut params: Vec<Box<dyn rusqlite::ToSql>> = Vec::new();

    if let Some(ref name) = request.name {
        updates.push("name = ?");
        params.push(Box::new(name.clone()));
    }

    if let Some(ref desc) = request.description {
        updates.push("description = ?");
        params.push(Box::new(desc.clone()));
    }

    if let Some(cover_id) = request.cover_media_id {
        updates.push("cover_media_id = ?");
        params.push(Box::new(cover_id));
    }

    if !updates.is_empty() {
        params.push(Box::new(request.album_id));
        let sql = format!("UPDATE albums SET {} WHERE id = ?", updates.join(", "));
        let param_refs: Vec<&dyn rusqlite::ToSql> = params.iter().map(|p| p.as_ref()).collect();
        execute_query(&conn, &sql, &param_refs)?;
    }

    let album = fetch_one(
        &conn,
        queries::albums::SELECT_WITH_COUNT,
        &[&request.album_id],
        map_album_row,
    )?
    .ok_or_else(|| AppError::Internal("Failed to update album".to_string()))?;

    Ok(Json(album))
}

async fn delete_album(
    State(state): State<AppState>,
    current_user: CurrentUser,
    Json(request): Json<AlbumDeleteRequest>,
) -> AppResult<Json<serde_json::Value>> {
    let conn = state.pool.get().map_err(AppError::Pool)?;

    let exists = fetch_one(
        &conn,
        queries::albums::CHECK_OWNERSHIP,
        &[&request.album_id, &current_user.id],
        |row| row.get::<_, i64>(0),
    )?;

    if exists.is_none() {
        return Err(AppError::NotFound("Album not found".to_string()));
    }

    execute_query(
        &conn,
        queries::albums::DELETE_ACCESS,
        &[&request.album_id, &current_user.id],
    )?;

    Ok(Json(
        serde_json::json!({"message": "Album deleted successfully"}),
    ))
}

async fn add_media_to_album(
    State(state): State<AppState>,
    current_user: CurrentUser,
    Json(request): Json<AlbumAddMediaRequest>,
) -> AppResult<Json<serde_json::Value>> {
    let conn = state.pool.get().map_err(AppError::Pool)?;

    let exists = fetch_one(
        &conn,
        queries::albums::CHECK_OWNERSHIP,
        &[&request.album_id, &current_user.id],
        |row| row.get::<_, i64>(0),
    )?;

    if exists.is_none() {
        return Err(AppError::NotFound("Album not found".to_string()));
    }

    let max_pos: i64 = fetch_one(
        &conn,
        queries::albums::SELECT_MAX_POSITION,
        &[&request.album_id],
        |row| row.get(0),
    )?
    .unwrap_or(-1);

    let mut next_pos = max_pos + 1;

    for media_id in &request.media_ids {
        let media_exists = fetch_one(
            &conn,
            queries::media::CHECK_EXISTS,
            &[media_id, &current_user.id],
            |row| row.get::<_, i64>(0),
        )?;

        if media_exists.is_none() {
            continue;
        }

        let _ = conn.execute(
            queries::albums::ADD_MEDIA,
            rusqlite::params![request.album_id, media_id, next_pos],
        );
        next_pos += 1;
    }

    Ok(Json(serde_json::json!({"message": "Media added to album"})))
}

async fn remove_media_from_album(
    State(state): State<AppState>,
    current_user: CurrentUser,
    Json(request): Json<AlbumRemoveMediaRequest>,
) -> AppResult<Json<serde_json::Value>> {
    let conn = state.pool.get().map_err(AppError::Pool)?;

    let exists = fetch_one(
        &conn,
        queries::albums::CHECK_OWNERSHIP,
        &[&request.album_id, &current_user.id],
        |row| row.get::<_, i64>(0),
    )?;

    if exists.is_none() {
        return Err(AppError::NotFound("Album not found".to_string()));
    }

    for media_id in &request.media_ids {
        conn.execute(
            queries::albums::REMOVE_MEDIA,
            rusqlite::params![request.album_id, media_id],
        )?;
    }

    Ok(Json(
        serde_json::json!({"message": "Media removed from album"}),
    ))
}

async fn list_albums(
    State(state): State<AppState>,
    current_user: CurrentUser,
) -> AppResult<Json<AlbumListResponse>> {
    let conn = state.pool.get().map_err(AppError::Pool)?;

    let albums = fetch_all(
        &conn,
        queries::albums::SELECT_ALL_FOR_USER,
        &[&current_user.id],
        map_album_row,
    )?;

    Ok(Json(AlbumListResponse { albums }))
}

async fn get_album(
    State(state): State<AppState>,
    current_user: CurrentUser,
    Json(request): Json<AlbumGetRequest>,
) -> AppResult<Json<AlbumDetailResponse>> {
    let conn = state.pool.get().map_err(AppError::Pool)?;

    let exists = fetch_one(
        &conn,
        queries::albums::CHECK_OWNERSHIP,
        &[&request.album_id, &current_user.id],
        |row| row.get::<_, i64>(0),
    )?;

    if exists.is_none() {
        return Err(AppError::NotFound("Album not found".to_string()));
    }

    let album = fetch_one(
        &conn,
        queries::albums::SELECT_BY_ID,
        &[&request.album_id],
        |row| {
            Ok(AlbumBasic {
                id: row.get(0)?,
                name: row.get(1)?,
                description: row.get(2)?,
                cover_media_id: row.get(3)?,
                created_at: row.get(5)?,
            })
        },
    )?
    .ok_or_else(|| AppError::NotFound("Album not found".to_string()))?;

    let media = fetch_all(
        &conn,
        queries::albums::SELECT_MEDIA,
        &[&request.album_id],
        map_media_row,
    )?;

    Ok(Json(AlbumDetailResponse {
        id: album.id,
        name: album.name,
        description: album.description,
        cover_media_id: album.cover_media_id,
        media,
        created_at: album.created_at,
    }))
}

async fn reorder_album_media(
    State(state): State<AppState>,
    current_user: CurrentUser,
    Json(request): Json<AlbumReorderRequest>,
) -> AppResult<Json<serde_json::Value>> {
    let conn = state.pool.get().map_err(AppError::Pool)?;

    let exists = fetch_one(
        &conn,
        queries::albums::CHECK_OWNERSHIP,
        &[&request.album_id, &current_user.id],
        |row| row.get::<_, i64>(0),
    )?;

    if exists.is_none() {
        return Err(AppError::NotFound("Album not found".to_string()));
    }

    for (i, media_id) in request.media_ids.iter().enumerate() {
        conn.execute(
            queries::albums::UPDATE_POSITION,
            rusqlite::params![i as i64, request.album_id, media_id],
        )?;
    }

    Ok(Json(
        serde_json::json!({"message": "Album reordered successfully"}),
    ))
}
